programs/drift-competitions/src/state/competition.rs

```diff
@@ -1,30 +1,409 @@
+
+    pub fn competitor_can_be_settled(&self, competitor: &Competitor) -> bool {
+        return self.round_number == competitor.competition_round_number;
+    }
+
+    pub fn settle_competitor(
+        &mut self,
+        competitor: &mut Competitor,
+        user_stats: &UserStats,
+        now: i64,
+    ) -> CompetitionResult {
+        self.validate_round_ready_for_settlement(now)?;
+
+        if !self.competitor_can_be_settled(competitor) {
+            return Ok(()); // gracefully skip/fail
+        }
+
+        if competitor.unclaimed_winnings == 0 {
+            let round_score = competitor.calculate_round_score(user_stats)?;
+
+            let round_score_capped = if self.max_entries_per_competitor > 0 {
+                round_score.min(self.max_entries_per_competitor)
+            } else {
+                round_score
+            };
+
+            let new_total_score_settled = self
+                .total_score_settled
+                .safe_add(round_score_capped.cast()?)?;
+
+            competitor.min_draw = self.total_score_settled;
+            competitor.max_draw = new_total_score_settled;
+            competitor.competition_round_number =
+                competitor.competition_round_number.safe_add(1)?;
+
+            self.total_score_settled = new_total_score_settled;
+        }
+
+        self.number_of_competitors_settled = self.number_of_competitors_settled.saturating_add(1);
+
+        Ok(())
+    }
+
+    pub fn calculate_sponsor_max_prize(
+        &self,
+        spot_market: &SpotMarket,
+        vault_balance: u64,
+    ) -> CompetitionResult<u64> {
+        let protocol_owned_shares = spot_market
+            .insurance_fund
+            .total_shares
+            .safe_sub(spot_market.insurance_fund.user_shares)?;
+
+        let protocol_owned_amount = if_shares_to_vault_amount(
+            protocol_owned_shares,
+            spot_market.insurance_fund.total_shares,
+            vault_balance,
+        )?;
+
+        let max_prize = protocol_owned_amount
+            .saturating_sub(self.sponsor_info.min_sponsor_amount)
+            .safe_mul(self.sponsor_info.max_sponsor_fraction)?
+            .safe_div(PERCENTAGE_PRECISION_U64)?;
+
+        Ok(max_prize)
+    }
+
+    pub fn calculate_prize_buckets_and_ratios(
+        self,
+        spot_market: &SpotMarket,
+        vault_balance: u64,
+    ) -> CompetitionResult<([u128; 3], Vec<u128>)> {
+        let max_prize: u128 = self
+            .calculate_sponsor_max_prize(spot_market, vault_balance)?
+            .cast()?;
+
+        // prize ratios match [$1k, $5k, $10k] ratios, but lower prizes never exceed 1k, 5k
+        let prize_buckets = [
+            (1000 * QUOTE_PRECISION).min(max_prize / 10),
+            (5000 * QUOTE_PRECISION).min(max_prize / 2),
+            max_prize,
+        ];
+
+        let total_prize_bucket: u128 = prize_buckets.iter().sum();
+        let mut ratios: Vec<u128> = vec![0; prize_buckets.len()]; // Using .len() to set the size
```

Comment:
why use the len? the fn signature already hardcodes it to 3

Reply:
i dont have to change it in two places